track_syscalls/main.go

```diff
@@ -17,129 +20,170 @@ import (
+	mode := flag.String("mode", "enforce", "Execution mode: 'build', 'enforce' or 'trace'")
 	flag.Parse()
 
-	// Check if the binary path is provided
 	if *binaryPath == "" {
 		log.Fatal("Please provide a binary path using the -binary flag")
 	}
-	if *allowlistPath == "" {
-		log.Fatal("Please provide a path to the allowlist JSON file using the -allowlist flag")
-	}
-	allowlist, err := loadAllowlist(*allowlistPath)
-	if err != nil {
-		log.Fatalf("loading allowlist: %v", err)
-	}
 
-	// Populate function symbols cache from the binary
-	if err := populateSymbolCache(*binaryPath); err != nil {
-		log.Fatalf("populating symbol cache: %v", err)
+	switch *mode {
+	case "build":
+		runBuildMode(*binaryPath)
+	case "enforce":
+		runEnforceMode(*binaryPath)
+	case "trace":
+		runTraceMode(*binaryPath)
+	default:
+		log.Fatalf("Invalid mode: %s. Use 'build', 'enforce' or 'trace'", *mode)
 	}
 
-	// Subscribe to signals for terminating the program.
-	stopper := make(chan os.Signal, 1)
-	signal.Notify(stopper, os.Interrupt, syscall.SIGTERM)
+}
 
+func setupEBPF() (*ebpfObjects, *ringbuf.Reader, *link.Link, error) {
 	// Allow the current process to lock memory for eBPF resources.
 	if err := rlimit.RemoveMemlock(); err != nil {
-		log.Fatal(err)
+		return nil, nil, nil, fmt.Errorf("removing memlock: %w", err)
 	}
 
 	// Load pre-compiled eBPF program and maps into the kernel.
 	objs := ebpfObjects{}
 	if err := loadEbpfObjects(&objs, nil); err != nil {
-		log.Fatalf("loading objects: %v", err)
+		return nil, nil, nil, fmt.Errorf("loading objects: %w", err)
 	}
-	defer objs.Close()
 
 	// Open a tracepoint and attach the pre-compiled program.
 	tp, err := link.Tracepoint("raw_syscalls", "sys_enter", objs.TraceSyscall, nil)
 	if err != nil {
-		log.Fatalf("opening tracepoint: %s", err)
+		objs.Close()
+		return nil, nil, nil, fmt.Errorf("opening tracepoint: %w", err)
 	}
-	defer tp.Close()
 
-	// Open a ringbuf reader from userspace RINGBUF map described in the
-	// eBPF C program.
+	// Open a ringbuf reader from userspace RINGBUF map described in the eBPF C program.
 	rd, err := ringbuf.NewReader(objs.Events)
 	if err != nil {
-		log.Fatalf("opening ringbuf reader: %s", err)
+		tp.Close()
+		objs.Close()
+		return nil, nil, nil, fmt.Errorf("opening ringbuf reader: %w", err)
 	}
-	defer rd.Close()
-
-	// Close the reader when the process receives a signal, which will exit
-	// the read loop.
-	go func() {
-		<-stopper
-
-		if err := rd.Close(); err != nil {
-			log.Fatalf("closing ringbuf reader: %s", err)
-		}
-	}()
 
-	log.Println("Waiting for events..")
+	return &objs, rd, &tp, nil
+}
 
-	// bpfEvent is generated by bpf2go.
-	var event ebpfEvent
-	for {
-		record, err := rd.Read()
-		if err != nil {
-			if errors.Is(err, ringbuf.ErrClosed) {
-				log.Println("Received signal, exiting..")
-				return
+func runBuildMode(binaryPath string) {
+	syscalls := make(map[string]map[int]bool)
```

Comment:
I am bit confused about what this `make` function does. Could you explain? :sweat_smile:

Reply:
`make` in Go create a slice, a dynamic data structure. 

We are creating a map of maps. The key of the outer map is the `callerPackage` name (string). For each `callerPackage`, we have an inner map, which the keys are syscall numbers (int). Each syscall number have an associated boolean value representing "allowed" or "denied". This structure allows for efficient allowlist check.